---
title: "01_R_studio_install_packages_basics"
author: "JR"
date: "7/20/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
library(GenomicRanges)
```

Today we are going to connect to Rstudio on the Fiji-viz server.

Accompanying Youtube Video:
https://youtu.be/2ESg9rpGQr8


We simply go here and login with our identikey:
https://fiji-viz.colorado.edu/
(click Rstudio button) -> login

# If you are taking this course elsewhere you can run most of the code on a local Rstudio (free)
https://www.rstudio.com/products/rstudio/download/

Please see video for an orientation of Rstudio:
https://www.youtube.com/watch?v=5SnLzPKa8bM

# In class we will login and make our new R project for class.
Please make sure you clone the class repo from git here:
/scratch/Shares/rinnclass/CLASS_2023/<Your folder>
# Pleae see this video for how to clone and connect github:
https://youtu.be/dTlz2sJgv1A


#################################
# 1 Make R project in Rstudio
#################################

Let's start by making an Rproject -- that will also connect to git.
In the top right hand corner there is a dropdown with "R shield"

(a) select open new project
(b) select from an existing folder (the class_2023 folder you cloned or forked)
(c) name the project.

#################################
# 2 Install packages
#################################

Now we will use Install.packages() for most packages. This will install any packages in CRAN repository. Let's try one.
```{R install packages}
?install.packages
install.packages("tidyverse")
```
Cool we have our first package installed.

However sometimes we need to install from Bioconductor and need to use BiocManager::install function
```{R install BiocManager}
if (!requireNamespace("BiocManager", quietly = TRUE))
install.packages("BiocManager", version = '3.10')
```

Let's now install genomic ranges package we will be using for most of class
# note we will have two classes on "Iranges" and Granges before we use GenomicRanges package
```{R install genomic ranges}
# Here is an example BiocManager::install
BiocManager::install("GenomicRanges")
# try installing all the packages listed below.
# note use either install.packages or BiocManager::install
```


################
# 3 {r setup, }
################
Note at the top of the document we source packages and functions we created.
The whole chunk looks like this:
```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
# this is providing information on how to "knit" the .RMD file.
# You can make really pretty results and corresponding code.
# Here is an example of a knitted document from last year:
# /scratch/Shares/rinnclass/CLASS_2022/EricY/bchm/exercises/analysis/final

options(stringsAsFactors = F)
# This is an important thing to have. When you read in a table R default is to
# make the columns have a "factor" associated with them. This is like a "meta variable"
# The col still has a name but that name is now an associated factor above the name.
# When we run RNAseq pipelines the comparison samples are "factorized" so the program
# knows which cols represent which samples for differential expression etc.

library(tidyverse)
library(GenomicRanges)
# This is how we load packages with "library(package X)

```


############################################
Exercise : install packages needed for class
############################################

# Hint google the package name and see if it is CRAN (install.packages)
# Or if in Bioconductor (BiocManager::install)

# probably worth reading what they do while installing...

Iranges
  ## Bioconductor 
  IRanges is a library for representing ranges of integers, which is useful in genomics, 
  because we have base pair ranges that we'd like to manipulate.There's a very detailed vignette on IRanges. 
  IRanges have a start, an end and a width.


ggplot2
  ## CRAN (Grmmar for graphics)
  ggplot2 is a plotting package that provides helpful commands to create complex plots from data in a data frame. 
  It provides a more programmatic interface for specifying what variables to plot, how they are displayed, 
  and general visual properties.
  
broom
  ## CRAN
  a package for tidying statistical models into data frames.
  the output of R statistical modeling functions isn’t tidy, meaning it’s difficult to manipulate and 
  recombine in downstream analyses and visualizations,
  Broom bridges the gap from untidy outputs of predictions and estimations to the tidy data we want to work with.
  
Gviz
  ## Bioconductor
  The Gviz package aims to provide a structured visualization framework to plot any type of data along genomic coordinates.
  It also allows to integrate publicly available genomic annotation data from sources like UCSC or ENSEMBL.
  
ggdendro
  ## CRAN
  ggdendrogram() is a wrapper around ggplot() to create a dendrogram using a single line of code.
  ggdendro can be used to extract the plot data from dendrogram and for drawing a dendrogram using ggplot2
  The height of the branch points indicates how similar or different they are from each other: the greater the height, 
  the greater the difference.
  We can use a dendrogram to represent the relationships between any kinds of entities as long
  as we can measure their similarity to each other.
  
ggrepel
  ## CRAN
  ggrepel provides geoms for ggplot2 to repel overlapping text labels: geom_text_repel() geom_label_repel()
  ggrepel implements functions to repel overlapping text labels away from each other 
  and away from the data points that they label
  
pheatmap
  ## CRAN (Pretty heatmaps)
  A function to draw clustered heatmaps where one has better control over some graphical parameters such as cell size, etc.
  
httr
  ## CRAN (Tools for Working with URLs and HTTP)
  The aim of httr is to provide a wrapper for the curl package, customised to the demands of modern web APIs.
  functions for the most important http verbs: GET(), HEAD(), PATCH(), PUT(), DELETE() and POST().
  Automatic connection sharing across requests to the same website (by default, curl handles are managed automatically),
  cookies are maintained across requests, and a up-to-date root-level SSL certificate store is used.
  Requests return a standard reponse object that captures the http status line, headers and body, 
  along with other useful information.
  
janitor
  ## CRAN
  janitor has simple functions for examining and cleaning dirty data. Janitor
  - perfectly format data.frame column names;
  - create and format frequency tables of one, two, or three variables - think an improved table();
  - provide other tools for cleaning and examining data.frames.
  
purrr
  ## CRAN
  purrr enhances R's functional programming (FP) toolkit by providing a complete and consistent set of tools 
  for working with functions and vectors.
  
DESeq2 
  ## Bioconductor - Differential gene expression analysis
  The DESeq2 package is designed for normalization, visualization, and differential analysis of high- dimensional count data
  It makes use of empirical Bayes techniques to estimate priors for log fold change and dispersion, 
  and to calculate posterior estimates for these quantities.
  
  ## DESeq and EdgeR are very similar and both assume that no genes are differentially expressed.
  DESeq uses a "geometric" normalisation strategy, whereas
  EdgeR is a weighted mean of log ratios-based method.
  Both normalise data initially via the calculation of size / normalisation factors.
  
  
ggpubr
  ## CRAN  - ggpubr: 'ggplot2' Based Publication Ready Plots
  ggpubr' provides some easy-to-use functions for creating and customizing 'ggplot2'- based publication ready plots


